### ISSUE TITLE
Suggestion: Profiler Improvements

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#832]
Hey, there. 

So I've been thinking about improving the Profiler a bit, pushing it more into Debug Mode territory. Here's a couple of ideas:

- Render the center points of GameObjects (like the physics visualizer, but works with non-static objects). The points could be rendered as different colors depending on if the object is running a custom class or has a `components` list size of greater than 0 (and so, is considered "active") or not.

- Drawing lines between parents and children.

- In a small pane, show the active scenes and their ordering.

- Mousing over an object shows a pane with the object's name. Clicking expands the pane, showing the object's position, scale, and rotation. Clicking again re-collapses the pane.

- Add shortcuts to enable and disable the profiler or physics visualization while in-game (say, `Tab` and `Ctrl+Tab`). This could be disabled on release by remapping the shortcuts to `null`.

- Future idea: render cameras' frustums.

**Comment by SolarLune** [comment]
A couple more ideas:

- Press a button over the object to reset it (call `init()` on it again).

- Timeline buttons that pause, single step through, and slow-down time.

- Right-click (?) to pick up an object. Right-click again to drop it (useful for moving objects around, testing out physics behaviors or level layouts quickly without having to drop back to Blender).

- With that in mind, a debug camera that you can switch to and from, controlled using standard "fly-mode" controls (this could just be a component that Blender's default cam uses).

**Comment by GoranM** [comment]
Well, it's something you should probably discuss with @rafcolson, but personally, I don't think a full on "debug mode" is worth the additional complexity.

If you need to focus on a particular data set, for debugging purposes, it doesn't really seem that difficult to do that with existing capabilities.

**Comment by SolarLune** [comment]
> I don't think a full on "debug mode" is worth the additional complexity.

Yeah, it would be additionally complex, but I don't think it would be too much additional code to implement (for some of these simpler ideas, anyway). I believe it'd be useful enough for general usage with most projects to include it with the engine.

I'll wait to see what @rafcolson thinks.

**Comment by rafcolson** [comment]
Okay, so here are my thoughts.

>- Render the center points of GameObjects (like the physics visualizer, but works with no-collision objects). The points could be rendered as different colors depending on if the object is running a custom class or has a components list size of greater than 0 (and so, is considered "active") or not.

I think such profiling preferences are to specific for the engine. But if anyone would require of such, it should be easy to add a `SceneVisualizer.java` component to an empty in the scene.

>- Drawing lines between parents and children.

>- Mousing over an object shows a pane with the object's name. Clicking expands the pane, showing the object's position, scale, and rotation. Clicking again re-collapses the pane.

Same thing, I think these functionalities could be part of `SceneVisualizer.java`.

>- In a small pane, show the active scenes and their ordering.

I think that this could be useful to add to the Profiler, with an option to disable or enable `scenesVisible`.

>- Add shortcuts to enable and disable the profiler or physics visualization while in-game (say, Tab and Ctrl+Tab). This could be disabled on release by remapping the shortcuts to null.

It would be fairly simple to manage this yourself if there's an option to enable and disable Physics Visualization in BDX. In my opinion there's no need for engine shortcuts if in-game shortcuts can be mapped externally rather easily.

>- Future idea: render cameras' frustums.

I'm not sure what you mean by that. Could you explain a bit?

>- Press a button over the object to reset it (call init() on it again).

To specific to add to internals, in my opinion. It could be part of a `SceneVisualizer.java` component though.

>- Timeline buttons that pause, single step through, and slow-down time.

Quite specific. Could be part of a `TimeController.java` component. Besides, for this to work internally, there should first be an option to dynamically change the speed of physics calculation relative to the tick rate, as the tick rate should remain 60, while slowing down time. Such option might not only serve debugging purposes, but also open doors for alternative game designs. So in the long run, it wouldn't have to be part of the Profiler, as it will be easy to manage time by altering `Bdx.speed` or something like that.

>- Right-click (?) to pick up an object. Right-click again to drop it (useful for moving objects around, testing out physics behaviors or level layouts quickly without having to drop back to Blender).

>- With that in mind, a debug camera that you can switch to and from, controlled using standard "fly-mode" controls (this could just be a component that Blender's default cam uses).

Both are very specific. Level testing and Fly mode could also be part of a component.

>Yeah, it would be additionally complex, but I don't think it would be to much additional code to implement (for some of these simpler ideas, anyway). I believe it'd be useful enough for general usage with most projects to include it with the engine.

It may not be to much additional code to implement, but I don't agree that these features embody general usage with most projects. It's my opinion that utilizing such features could help but are far from necessary with most projects. For example, physics behavior of objects in BDX should be predictable enough so "Level Testing" wouldn't be required. "Scene Visualization" would be useful with complex or specific scenes.

In conclusion, I think the Profiler could show info about active scenes. That's my opinion, but it's up to @GoranM to decide.

>If you need to focus on a particular data set, for debugging purposes, it doesn't really seem that difficult to do that with existing capabilities.

I have to agree.

**Comment by SolarLune** [comment]
Cool, thanks for the feedback.

> Render center points

> I think such profiling preferences are to specific for the engine.

I'm not sure what you mean by "profiling preferences" - do you mean the color of the center points, or that they're colored at all, or what? 
____________

Rendering center points of GameObjects can be useful, since rendering the center points of objects using physics visualization:

1) Renders a lot of "crap data" in the form of the mesh faces, if you don't need them, and
2) Renders center points only for solid objects, which means non-solid objects aren't visualized.

> I'm not sure what you mean by that. Could you explain a bit?

I meant a method to draw cameras' frustums' limits in-engine (since they're not visualized in BDX) - a means to see what the camera can view without switching to it. That could be part of that scene visualization component. However:

> Mousing over an object shows a pane with the object's name. Clicking expands the pane, showing the object's position, scale, and rotation. Clicking again re-collapses the pane.

> Same thing, I think these functionalities could be part of SceneVisualizer.java.

I'm not sure how this would be able to be added in a single component (for BDX or standalone) since it would require preparation of a Text object and a plane (or using existing ones) to serve as the background panel.

> It would be fairly simple to manage this yourself if there's an option to enable and disable Physics Visualization in BDX. In my opinion there's no need for engine shortcuts if in-game shortcuts can be mapped externally rather easily.

That's fine. I'm not sure if it's possible to enable and disable the profiler and physics visualization live in-game, though. I'll have to check later.

> Pausing and resuming

> Quite specific. Could be part of a TimeController.java component. Besides, for this to work internally, there should first be an option to dynamically change the speed of physics calculation relative to the tick rate, as the tick rate should remain 60, while slowing down time. Such option might not only serve debugging purposes, but also open doors for alternative game designs. So in the long run, it wouldn't have to be part of the Profiler, as it will be easy to manage time by altering Bdx.speed or something like that.

Hmm, I meant pausing and resuming via buttons onscreen - I don't really see how it would be specific, since many different games and genres could use this debugging feature. I do agree, though, that that option would be nice for gameplay opportunities, as well, and would be generally better than my idea for implementation (pausing and resuming scenes continuously).
_____

So to sum up, the reason I'm proposing these things is because I think they'd be pretty useful at higher-level, more complex development and some features require engine access to work or be shared easily amongst projects. I think I could definitely use a few with Gearend and would like to share the results with BDX master. Not every feature could be included, which is fine - these are all just ideas.

**Comment by rafcolson** [comment]
>I'm not sure what you mean by "profiling preferences" - do you mean the color of the center points, or that they're colored at all, or what?

I meant both.

>Rendering center points of GameObjects can be useful, since rendering the center points of objects using physics visualization:

>1) Renders a lot of "crap data" in the form of the mesh faces, if you don't need them, and
2) Renders center points only for solid objects, which means non-solid objects aren't visualized.

Wouldn't you agree that rendering center points would only be useful in specific situations? Not to mention a color system to indicate the use of classes and components.

Furthermore, such preferences seem to be more or less related to Physics Visualization, so why not optimize this instead. Add center points of NO_COLLISION objects and an option to exclude GameObjects.

>I'm not sure how this would be able to be added in a single component (for BDX or standalone) since it would require preparation of a Text object and a plane (or using existing ones) to serve as the background panel.

I was thinking of utilizing `Bdx.profiler.props` for this. However, `Bdx.profiler.props` would have to be a `LinkedHashMap` instead of just a `HashMap`. It may also be necessary to extend the number of characters for keys (could be done dynamically) and add support for text lines. By improving `Bdx.profiler.props` you wouldn't have to create a new system.

That said. It might be neat to put `Bdx.profiler.props` in a separate pane (top right for example).

>That's fine. I'm not sure if it's possible to enable and disable the profiler and physics visualization live in-game, though. I'll have to check later.

For the profiler, I think it would be better to put `__Profiler.bdx` in the `gen` folder once it is exported and export regardless of the *Show Framerate and Profile* being enabled or not.

The "physviz" parameter in line `world.setDebugDrawer(new Bullet.DebugDrawer(json.get("physviz").asBoolean()));` in `Bdx.java`, will influence the private `debug` field. Toggling this boolean might disable/enable Physics Visualization, although I'm not sure. Didn't test it.

>Not every feature could be included, which is fine - these are all just ideas.

My thoughts are also just thoughts. My main point is that providing the necessary infrastructure to be able to implement advanced profiling systems oneself, would be the best of both worlds. Improve `Bdx.profiler.props` and add a reference `Scene.debugDrawer`, to be able to toggle `debug`. But it's up to @GoranM really.

PS: Enabling/disabling visualization of specific objects would require extending `DiscreteDynamicsWorld`. Filtering `CollisionWorld.collisionObjects` could only be done by overriding `DiscreteDynamicsWorld.debugDrawWorld()`. See [here](http://jbullet.advel.cz/javadoc/com/bulletphysics/dynamics/DiscreteDynamicsWorld.html).

**Comment by GoranM** [comment]
> My main point is that providing the necessary infrastructure to be able to implement advanced profiling systems oneself, would be the best of both worlds.

I share the same view.

**Comment by SolarLune** [comment]
> Wouldn't you agree that rendering center points would only be useful in specific situations? Not to mention a color system to indicate the use of classes and components? I can be wrong.

Not really, no. Anytime you would want to display where objects are and not display the physics mesh (which, for me, is sometimes), this would be useful. One of the main problems is that all static geometry is physically visualized, which can be slow and visually distracting when you're looking for something in particular and you have a lot of level geometry.

But it's fine, we can look into a means to allow the user to more easily tweak how objects are visualized through the physics visualization.

> That said. It might be neat to reposition Bdx.profiler.props in a separate pane (top right for example).

I'd be cool with that.
